File: modules/cyc-bridge/ios/CycBridgeModule.swift
import ExpoModulesCore
import Foundation

private final class BridgeException: Exception {
  private let bridgeCode: String
  private let message: String
  init(code: String, message: String) {
    bridgeCode = code
    self.message = message
    super.init()
  }
  override var code: String { bridgeCode }
  override var reason: String { message }
}

private func rejectRead(_ promise: Promise, _ error: Error) {
  if let storage = error as? PowerLogStorageError, let code = storage.bridgeCode {
    promise.reject(BridgeException(code: code, message: storage.localizedDescription))
  } else if error is CycError || error is WorkoutDataError {
    promise.reject(BridgeException(code: "E_WORKOUT", message: error.localizedDescription))
  } else { promise.reject(error) }
}

struct CatalogPageOptions: Record {
  @Field var limit: Int = 100
  @Field var beforeStartedAt: String?
  @Field var beforeID: String = ""
}

struct CycConnectOptions: Record {
  @Field var deviceId: String = ""
  @Field var hz: Double = 2
}

struct WorkoutStartOptions: Record {
  @Field var indoor: Bool = false
  @Field var useWatch: Bool = true
  @Field var saveToHealth: Bool = true
  @Field var recordGPS: Bool?
  @Field var sampleHz: Double?
}

struct MonitorAnchorOptions: Record {
  @Field var metric: String = ""
  @Field var observationId: String = ""
  var anchor: MonitorObservationAnchor { MonitorObservationAnchor(metric: metric, observationId: observationId) }
}

struct MonitorReadOptions: Record {
  @Field var source: String = "live"
  @Field var id: String?
  @Field var generation: Int = 0
  @Field var expectedRevision: String?
  @Field var sinceRevision: String?
  @Field var startSeconds: Double?
  @Field var endSeconds: Double?
  @Field var metrics: [String] = []
  @Field var buckets: Int = 128
  @Field var pixelWidth: Double?
  @Field var seconds: Double?
  @Field var includeEndpoints: Bool = false
  @Field var anchor: MonitorAnchorOptions?
  @Field var startAnchor: MonitorAnchorOptions?
  @Field var endAnchor: MonitorAnchorOptions?
  @Field var distanceSource: String = "auto"
  var request: MonitorRequest {
    MonitorRequest(source: source, id: id, generation: generation, expectedRevision: expectedRevision,
      sinceRevision: sinceRevision, startSeconds: startSeconds, endSeconds: endSeconds, seconds: seconds,
      metrics: metrics, buckets: buckets, pixelWidth: pixelWidth, includeEndpoints: includeEndpoints,
      anchor: anchor?.anchor, startAnchor: startAnchor?.anchor, endAnchor: endAnchor?.anchor, distanceSource: distanceSource)
  }
}

public final class CycBridgeModule: Module {
  private let observerID = UUID()

  public func definition() -> ModuleDefinition {
    let engine = CycEngine.shared
    Name("CycBridge")
    Events("onDevice", "onState", "onSample", "onWorkoutState")

    View(MonitorRasterView.self) {
      Events("onRenderStatus")
      Prop("sourceId") { (view: MonitorRasterView, value: String) in view.setSourceID(value) }
      Prop("scene") { (view: MonitorRasterView, value: String) in view.setScene(value) }
      Prop("sceneKey") { (view: MonitorRasterView, value: String) in view.setSceneKey(value) }
      Prop("presentation") { (view: MonitorRasterView, value: [Double]) in view.setPresentation(value) }
      Prop("selection") { (view: MonitorRasterView, value: String) in view.setSelection(value) }
      Prop("selectionTarget") { (view: MonitorRasterView, value: String) in view.setSelectionTarget(value) }
    }
    Function("getMonitorRasterDiagnostics") { MonitorRasterWorker.shared.diagnostics() }
    Function("resetMonitorRasterDiagnostics") { MonitorRasterWorker.shared.diagnostics(reset: true) }
    AsyncFunction("startMonitorRasterFrameDiagnostics") { MonitorRasterFrameDiagnostics.shared.start() }.runOnQueue(.main)
    AsyncFunction("getMonitorRasterFrameDiagnostics") { MonitorRasterFrameDiagnostics.shared.snapshot() }.runOnQueue(.main)
    AsyncFunction("stopMonitorRasterFrameDiagnostics") { MonitorRasterFrameDiagnostics.shared.stop() }.runOnQueue(.main)

    AsyncFunction("describeMonitorSource") { (options: MonitorReadOptions, promise: Promise) in
      MonitorDataStore.shared.submit(.describe, request: options.request) { result in
        switch result {
        case .success(let value): promise.resolve(value)
        case .failure(let error): rejectRead(promise, error)
        }
      }
    }
    AsyncFunction("readMonitorPlot") { (options: MonitorReadOptions, promise: Promise) in
      MonitorDataStore.shared.submit(.plot, request: options.request) { result in
        switch result {
        case .success(let value): promise.resolve(value)
        case .failure(let error): rejectRead(promise, error)
        }
      }
    }
    AsyncFunction("readMonitorLatest") { (options: MonitorReadOptions, promise: Promise) in
      MonitorDataStore.shared.submit(.latest, request: options.request) { result in
        switch result {
        case .success(let value): promise.resolve(value)
        case .failure(let error): rejectRead(promise, error)
        }
      }
    }
    AsyncFunction("inspectMonitorAt") { (options: MonitorReadOptions, promise: Promise) in
      MonitorDataStore.shared.submit(.inspect, request: options.request) { result in
        switch result {
        case .success(let value): promise.resolve(value)
        case .failure(let error): rejectRead(promise, error)
        }
      }
    }
    AsyncFunction("readMonitorRangeStats") { (options: MonitorReadOptions, promise: Promise) in
      MonitorDataStore.shared.submit(.stats, request: options.request) { result in
        switch result {
        case .success(let value): promise.resolve(value)
        case .failure(let error): rejectRead(promise, error)
        }
      }
    }
    AsyncFunction("monitorChangesSince") { (options: MonitorReadOptions, promise: Promise) in
      MonitorDataStore.shared.submit(.changes, request: options.request) { result in
        switch result {
        case .success(let value): promise.resolve(value)
        case .failure(let error): rejectRead(promise, error)
        }
      }
    }

    OnCreate { [weak self] in
      guard let self else { return }
      let id = self.observerID
      if #available(iOS 26.0, *) {
        let workout = WorkoutEngine.shared
        workout.queue.async { [weak self] in
          workout.addSink(id: id) { [weak self] body in
            DispatchQueue.main.async { [weak self] in self?.sendEvent("onWorkoutState", body) }
          }
        }
      }
      engine.queue.async { [weak self] in
        engine.addSink(id: id) { [weak self] event, body in
          // Events carry copies; Bluetooth and disk work stay on the native serial queue.
          DispatchQueue.main.async { [weak self] in self?.sendEvent(event, body) }
        }
      }
    }
    OnDestroy { [weak self] in
      guard let id = self?.observerID else { return }
      engine.queue.async { engine.removeSink(id: id) }
      if #available(iOS 26.0, *) {
        let workout = WorkoutEngine.shared
        workout.queue.async { workout.removeSink(id: id) }
      }
    }

    AsyncFunction("getState") { () -> [String: Any] in engine.state() }.runOnQueue(engine.queue)
    AsyncFunction("getDiagnostics") { () -> [String: Any] in engine.diagnostics() }.runOnQueue(engine.queue)
    AsyncFunction("readDiagnostics") { (promise: Promise) in
      engine.readDiagnostics { result in
        switch result {
        case .success(let jsonl): promise.resolve(jsonl)
        case .failure(let error): promise.reject(error)
        }
      }
    }.runOnQueue(engine.queue)
    AsyncFunction("startScan") { () throws in try engine.startScan() }.runOnQueue(engine.queue)
    AsyncFunction("stopScan") { () in engine.stopScan() }.runOnQueue(engine.queue)
    AsyncFunction("connect") { (options: CycConnectOptions) throws in
      try engine.connect(deviceID: options.deviceId, rate: options.hz)
    }.runOnQueue(engine.queue)
    AsyncFunction("disconnect") { () throws in try engine.disconnect() }.runOnQueue(engine.queue)

    AsyncFunction("getWorkoutState") { (promise: Promise) in
      if #available(iOS 26.0, *) {
        let workout = WorkoutEngine.shared
        workout.queue.async { promise.resolve(workout.state()) }
      } else {
        promise.resolve(["supported": false, "phase": "unavailable", "error": "Complete workout recording requires iOS 26 or newer."])
      }
    }
    AsyncFunction("getWorkoutPermissions") { (promise: Promise) in
      if #available(iOS 26.0, *) {
        WorkoutEngine.shared.getPermissions { result in
          switch result { case .success(let state): promise.resolve(state); case .failure(let error): promise.reject(error) }
        }
      } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    AsyncFunction("requestWorkoutPermissionsForOptions") { (options: WorkoutStartOptions, promise: Promise) in
      if #available(iOS 26.0, *) {
        WorkoutEngine.shared.requestPermissions(indoor: options.indoor, useWatch: options.useWatch, saveToHealth: options.saveToHealth, recordGPS: options.recordGPS) { result in
          switch result { case .success(let state): promise.resolve(state); case .failure(let error): promise.reject(error) }
        }
      } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    AsyncFunction("requestWorkoutPermissions") { (promise: Promise) in
      if #available(iOS 26.0, *) {
        WorkoutEngine.shared.requestPermissions { result in
          switch result { case .success(let state): promise.resolve(state); case .failure(let error): promise.reject(error) }
        }
      } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    AsyncFunction("startWorkout") { (options: WorkoutStartOptions, promise: Promise) in
      if #available(iOS 26.0, *) {
        let workout = WorkoutEngine.shared
        workout.queue.async {
          workout.start(indoor: options.indoor, useWatch: options.useWatch, saveToHealth: options.saveToHealth, recordGPS: options.recordGPS, sampleHz: options.sampleHz) { result in
            switch result { case .success(let state): promise.resolve(state); case .failure(let error): promise.reject(error) }
          }
        }
      } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    AsyncFunction("recoverWorkout") { (id: String, promise: Promise) in
      if #available(iOS 26.0, *) {
        let workout = WorkoutEngine.shared
        workout.queue.async { workout.recover(id) { result in
          switch result { case .success(let state): promise.resolve(state); case .failure(let error): promise.reject(error) }
        } }
      } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    AsyncFunction("pauseWorkout") { (id: String?, promise: Promise) in
      if #available(iOS 26.0, *) { self.workoutAction(promise) { try $0.pause(expectedID: id) } } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    AsyncFunction("resumeWorkout") { (id: String?, promise: Promise) in
      if #available(iOS 26.0, *) { self.workoutAction(promise) { try $0.resume(expectedID: id) } } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    AsyncFunction("markWorkoutLap") { (id: String?, promise: Promise) in
      if #available(iOS 26.0, *) { self.workoutAction(promise) { try $0.lap(expectedID: id) } } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    AsyncFunction("stopWorkout") { (id: String?, promise: Promise) in
      if #available(iOS 26.0, *) { self.workoutAction(promise) { try $0.stop(expectedID: id) } } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    AsyncFunction("discardWorkout") { (id: String, promise: Promise) in
      if #available(iOS 26.0, *) { self.workoutAction(promise) { try $0.stop(expectedID: id, discard: true) } } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    AsyncFunction("deleteWorkout") { (id: String, promise: Promise) in
      if #available(iOS 26.0, *) { self.workoutAction(promise) { try $0.deleteWorkout(id) } } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    AsyncFunction("listWorkouts") { (options: CatalogPageOptions?, promise: Promise) in
      if #available(iOS 26.0, *) { self.workoutAction(promise) { try $0.list(limit: options?.limit ?? 100, beforeStartedAt: options?.beforeStartedAt, beforeID: options?.beforeID ?? "") } } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    #if DEBUG
    AsyncFunction("addExampleRides") { (promise: Promise) in
      if #available(iOS 26.0, *) { self.workoutAction(promise) { try $0.addExampleRides() } } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    #endif
    AsyncFunction("readWorkout") { (id: String, distanceSource: String?, promise: Promise) in
      if #available(iOS 26.0, *) { self.workoutFileAction(promise) { try $0.read(id, distanceSource: distanceSource ?? "auto") } } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    AsyncFunction("exportWorkoutArchive") { (id: String, promise: Promise) in
      if #available(iOS 26.0, *) { self.workoutFileAction(promise) { try $0.exportOriginal(id) } } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
    AsyncFunction("exportWorkout") { (id: String, distanceSource: String?, promise: Promise) in
      if #available(iOS 26.0, *) { self.workoutFileAction(promise) { try $0.export(id, distanceSource: distanceSource ?? "auto") } } else { promise.reject(CycError.invalid("Complete workout recording requires iOS 26 or newer.")) }
    }
  }

  @available(iOS 26.0, *)
  private func workoutAction(_ promise: Promise, action: @escaping (WorkoutEngine) throws -> Any) {
    let engine = WorkoutEngine.shared
    engine.queue.async { do { promise.resolve(try action(engine)) } catch { rejectRead(promise, error) } }
  }

  @available(iOS 26.0, *)
  private func workoutFileAction(_ promise: Promise, action: @escaping (WorkoutEngine) throws -> Any) {
    let engine = WorkoutEngine.shared
    let deadline = ProcessInfo.processInfo.systemUptime + 60
    func attempt() {
      engine.fileQueue.async {
        do { promise.resolve(try action(engine)) }
        catch WorkoutDistanceError.pending where ProcessInfo.processInfo.systemUptime < deadline {
          engine.fileQueue.asyncAfter(deadline: .now() + 0.25) { attempt() }
        } catch WorkoutDistanceError.expired where ProcessInfo.processInfo.systemUptime < deadline {
          engine.fileQueue.asyncAfter(deadline: .now() + 0.25) { attempt() }
        } catch { rejectRead(promise, error) }
      }
    }
    attempt()
  }
}

File: modules/cyc-bridge/android/src/main/java/app/powerlog/bridge/CycBridgeModule.kt
package app.powerlog.bridge

import android.Manifest
import android.content.Intent
import android.net.Uri
import android.os.Build
import androidx.core.content.FileProvider
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File

class CycBridgeModule : Module() {
    private var healthContinuation: (() -> Unit)? = null
    private val engine
        get() = RecordingEngine.get(requireNotNull(appContext.reactContext))

    private val listener: (String, Payload) -> Unit = { name, value -> sendEvent(name, value) }

    private fun submit(promise: Promise, read: Boolean = false, body: () -> Any?) {
        val job = Runnable {
            try {
                engine.awaitReady()
                promise.resolve(body())
            } catch (error: Exception) {
                promise.reject("E_POWER_LOG", error.message ?: "Power Log operation failed", error)
            }
        }
        if (read) engine.reads.execute(job) else engine.handler.post(job)
    }

    private fun permissions(
        gps: Boolean,
        notifications: Boolean,
        promise: Promise,
        after: () -> Any?,
    ) {
        val required = buildList {
            if (Build.VERSION.SDK_INT >= 31) {
                add(Manifest.permission.BLUETOOTH_SCAN)
                add(Manifest.permission.BLUETOOTH_CONNECT)
            }
            if (gps || Build.VERSION.SDK_INT < 31) {
                add(Manifest.permission.ACCESS_COARSE_LOCATION)
                add(Manifest.permission.ACCESS_FINE_LOCATION)
            }
            if (notifications && Build.VERSION.SDK_INT >= 33)
                add(Manifest.permission.POST_NOTIFICATIONS)
        }
        val manager =
            appContext.permissions
                ?: run {
                    promise.reject("E_PERMISSIONS", "Permissions are unavailable", null)
                    return
                }
        if (gps)
            appContext.reactContext
                ?.getSharedPreferences("permissions", 0)
                ?.edit()
                ?.putBoolean("locationRequested", true)
                ?.apply()
        manager.askForPermissions({ submit(promise, body = after) }, *required.toTypedArray())
    }

    private fun ridePermissions(options: Payload, promise: Promise, after: () -> Any?) {
        val gps = options.flag("recordGPS", !options.flag("indoor"))
        if (!options.flag("saveToHealth")) {
            permissions(gps, true, promise, after)
            return
        }
        engine.reads.execute {
            try {
                engine.awaitReady()
                check(engine.health.available) { "Health Connect is unavailable." }
                val required = engine.health.permissions(gps)
                if (engine.health.granted().containsAll(required)) {
                    permissions(gps, true, promise, after)
                    return@execute
                }
                val activity = appContext.throwingActivity
                activity.runOnUiThread {
                    if (healthContinuation != null) {
                        promise.reject(
                            "E_PERMISSIONS",
                            "A permission request is already open",
                            null,
                        )
                        return@runOnUiThread
                    }
                    healthContinuation = { permissions(gps, true, promise, after) }
                    try {
                        activity.startActivityForResult(
                            Intent(activity, HealthPermissionsActivity::class.java)
                                .putExtra("gps", gps),
                            8642,
                        )
                    } catch (error: Exception) {
                        healthContinuation = null
                        promise.reject("E_PERMISSIONS", error.message, error)
                    }
                }
            } catch (error: Exception) {
                promise.reject("E_PERMISSIONS", error.message, error)
            }
        }
    }

    override fun definition() = ModuleDefinition {
        Name("CycBridge")
        Events("onDevice", "onState", "onSample", "onWorkoutState")
        OnCreate { engine.listeners.add(listener) }
        OnDestroy {
            engine.listeners.remove(listener)
            healthContinuation = null
        }
        OnActivityResult { _, (requestCode, _, _) ->
            if (requestCode == 8642) {
                val continuation = healthContinuation
                healthContinuation = null
                continuation?.invoke()
            }
        }
        View(MonitorRasterView::class) {
            Events("onRenderStatus")
            Prop("sourceId") { view: MonitorRasterView, value: String -> view.source(value) }
            Prop("sceneKey") { view: MonitorRasterView, value: String -> view.key(value) }
            Prop("scene") { view: MonitorRasterView, value: String -> view.scene(value) }
            Prop("presentation") { view: MonitorRasterView, value: List<Double> ->
                view.presentation(value)
            }
            Prop("selection") { view: MonitorRasterView, value: String -> view.selection(value) }
            Prop("selectionTarget") { view: MonitorRasterView, value: String -> view.target(value) }
        }
        AsyncFunction("getState") { promise: Promise -> submit(promise) { engine.bluetooth.state } }
        AsyncFunction("getDiagnostics") { promise: Promise ->
            submit(promise) { engine.bluetooth.diagnostics() }
        }
        AsyncFunction("readDiagnostics") { promise: Promise ->
            submit(promise) { json(engine.bluetooth.diagnostics()).toString() }
        }
        AsyncFunction("startScan") { promise: Promise ->
            permissions(false, false, promise) {
                engine.bluetooth.startScan()
                null
            }
        }
        AsyncFunction("stopScan") { promise: Promise ->
            submit(promise) {
                engine.bluetooth.stopScan()
                null
            }
        }
        AsyncFunction("connect") { options: Map<String, Any?>, promise: Promise ->
            engine.handler.post {
                try {
                    engine.awaitReady()
                    engine.checkBike(options.str("deviceId"))
                    engine.bluetooth.connect(
                        options.str("deviceId"),
                        options.num("hz", 2.0).toInt(),
                    ) { error ->
                        if (error == null) promise.resolve(null)
                        else promise.reject("E_BLUETOOTH", error.message, error)
                    }
                } catch (error: Exception) {
                    promise.reject("E_BLUETOOTH", error.message, error)
                }
            }
        }
        AsyncFunction("disconnect") { promise: Promise ->
            submit(promise) {
                engine.bluetooth.disconnect()
                null
            }
        }
        AsyncFunction("getWorkoutState") { promise: Promise -> submit(promise) { engine.state() } }
        AsyncFunction("getWorkoutPermissions") { promise: Promise ->
            submit(promise) { engine.permissionStatus() }
        }
        AsyncFunction("requestWorkoutPermissions") { promise: Promise ->
            permissions(true, true, promise) { engine.permissionStatus() }
        }
        AsyncFunction("requestWorkoutPermissionsForOptions") {
            options: Map<String, Any?>,
            promise: Promise ->
            ridePermissions(options, promise) { engine.permissionStatus() }
        }
        AsyncFunction("startWorkout") { options: Map<String, Any?>, promise: Promise ->
            ridePermissions(options, promise) { engine.start(options) }
        }
        AsyncFunction("pauseWorkout") { id: String?, promise: Promise ->
            submit(promise) { engine.action("pause", id) }
        }
        AsyncFunction("resumeWorkout") { id: String?, promise: Promise ->
            submit(promise) { engine.action("resume", id) }
        }
        AsyncFunction("markWorkoutLap") { id: String?, promise: Promise ->
            submit(promise) { engine.action("lap", id) }
        }
        AsyncFunction("stopWorkout") { id: String?, promise: Promise ->
            submit(promise) { engine.action("stop", id) }
        }
        AsyncFunction("discardWorkout") { id: String, promise: Promise ->
            submit(promise) { engine.action("discard", id) }
        }
        AsyncFunction("deleteWorkout") { id: String, promise: Promise ->
            submit(promise) { engine.delete(id) }
        }
        AsyncFunction("recoverWorkout") { id: String, promise: Promise ->
            submit(promise) { engine.recover(id) }
        }
        AsyncFunction("listWorkouts") { options: Map<String, Any?>, promise: Promise ->
            submit(promise, true) { engine.store.list(options) }
        }
        AsyncFunction("readWorkout") { id: String, source: String?, promise: Promise ->
            submit(promise, true) {
                RideExport(engine.context, engine.store, engine.distance, engine.monitor)
                    .detail(id, source ?: "auto")
            }
        }
        AsyncFunction("exportWorkout") { id: String, source: String?, promise: Promise ->
            submit(promise, true) {
                engine.store.withSavedRide(id) {
                    RideExport(engine.context, engine.store, engine.distance, engine.monitor)
                        .fit(id, source ?: "auto")
                }
            }
        }
        AsyncFunction("exportWorkoutArchive") { id: String, promise: Promise ->
            submit(promise, true) {
                engine.store.withSavedRide(id) {
                    RideExport(engine.context, engine.store, engine.distance, engine.monitor)
                        .archive(id)
                }
            }
        }
        listOf(
                "describeMonitorSource" to "describe",
                "readMonitorLatest" to "latest",
                "readMonitorPlot" to "plot",
                "inspectMonitorAt" to "inspect",
                "readMonitorRangeStats" to "stats",
                "monitorChangesSince" to "changes",
            )
            .forEach { (name, kind) ->
                AsyncFunction(name) { options: Map<String, Any?>, promise: Promise ->
                    submit(promise, true) {
                        engine.monitor.query(kind, engine.source(options), options)
                    }
                }
            }
        AsyncFunction("shareFile") { uri: String, promise: Promise ->
            try {
                val context = requireNotNull(appContext.reactContext)
                val source = File(requireNotNull(Uri.parse(uri).path)).canonicalFile
                val cache = context.cacheDir.canonicalFile
                require(source.path.startsWith(cache.path + File.separator) && source.isFile) {
                    "Export file is unavailable."
                }
                val exported = File(File(cache, "exports").apply { mkdirs() }, source.name)
                if (source != exported) source.copyTo(exported, true)
                val shared =
                    FileProvider.getUriForFile(
                        context,
                        "${context.packageName}.powerlog.files",
                        exported,
                    )
                val intent =
                    Intent(Intent.ACTION_SEND)
                        .setType(
                            if (exported.extension == "csv") "text/csv"
                            else "application/octet-stream"
                        )
                        .putExtra(Intent.EXTRA_STREAM, shared)
                        .addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                requireNotNull(appContext.currentActivity)
                    .startActivity(Intent.createChooser(intent, "Export ride"))
                promise.resolve(null)
            } catch (error: Exception) {
                promise.reject("E_EXPORT", error.message, error)
            }
        }
    }
}
